### ISSUE TITLE
provide permissions accessor for filepath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4813]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#4813|
|Type|enhancement|
|Created|2011-01-20 18:31:18Z|
|Branch|https://github.com/twisted/twisted/tree/provide-filepath-permissions-4813-2|

[#4712](https://github.com/twisted/twisted/issues/4712) provided a bunch of `statinfo` accessors, but is missing some kind of accessor for the permissions, which is needed by `twisted.protocols.ftp`.

<details><summary>Searchable metadata</summary>

```
trac-id__4813 4813
type__enhancement enhancement
reporter__cyli cyli
priority__normal normal
milestone__ 
branch__branches_provide_filepath_permissions_4813_2 branches/provide-filepath-permissions-4813-2
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1295548278000000 1295548278000000
changetime__1309563469000000 1309563469000000
version__None None
owner__cyli cyli
cc__exarkun cc__cyli cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli commented|
|-|-|

(In [30541]) Branching to 'provide-filepath-permissions-4813'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

(In [30542]) Add mode accessor to access file permissions.  Also added test.

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_3"></a>|@cyli removed owner|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/provide-filepath-permissions-4813)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @cyli|
|-|-|

I feel like `getMode` is a bit too broad of a way to expose this functionality.

First, it's partially redundant with other methods - `isdir` and `isfile`.  Presumably if we added `isSocket` or `isBlockDevice` it would become even more redundant.  Second, there's no way to extract useful information from the resulting integer without using the stdlib `stat` module.

I realize that this was in support of FTP, and that this will probably involve more changes to the FTP code.  While I realize maintaining that stuff is kind of a pain, the main line that this is in support of is
```
#!py
return ''.join([mode & (256 >> n) and 'rwx'[n % 3] or '-' for n in range(9)])
```
and I wouldn't mind seeing some code like _that_ get thrown away.

Also, `ZipPath` should have enough information to satisfy an FTP server, and synthesizing a mode integer would be trickier than creating some kind of structured object.  I'd like an interface like `path.permissions.user.read`, `path.permissions.group.execute` so that clients can make logical use of the object directly rather than indirectly via `stat.S_WHATEVER`.

I see that the FTP public interface does contain a trace of this bogus spec.  The `IFTPShell.list` method says that `permissions` is a "bitvector" and doesn't go into much detail.  But, I think that the name 'permissions' is actually misleading; for it to properly be permission information (the same thing conveyed by the actual protocol implementation) it would need to just be `stat.S_IRWXU | stat.S_IRWXG | stat.S_IRWXO`.  So, when FTP does get fixed up, it should first be fixed up to generate this number from these fields, and later be fixed to have a better `list` interface method which returns an object rather than a list of tuples of lists (or whatever that is).

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [31027]) Branching to 'provide-filepath-permissions-4813-2'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [31108]) Added Permissions class for filepath

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

(In [31111]) Provide __str__ methods for Permissions class

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

(In [31116]) Add a permissions accessor for FilePath

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli commented|
|-|-|

(In [31120]) Fixed bug where if other permissions > user or group permissions, it would interpret the bitmask incorrectly (e.g. for numbers like 001)

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

(In [31125]) Add isBlockDevice and isSocket methods

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_12"></a>|@cyli commented|
|-|-|

(In [31127]) Fix up some bugs

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_13"></a>|@cyli commented|
|-|-|

(In [31128]) Newsfile

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_14"></a>|@cyli commented|
|-|-|

(In [31132]) Python 2.4 doesn't support "X if C else Y" formations I guess

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_15"></a>|@cyli commented|
|-|-|

(In [31133]) Add @since markers

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_16"></a>|@cyli commented|
|-|-|

Redid the branch to follow some of glyph's feedback instead.  There is a Permissions class, and the permissions can be accessed via a getPermissions() method, rather than an attribute, so that caching the .permissions attribute isn't an issue.

Also put in the isBlockDevice and isSocket methods.

[seems to build ok](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/provide-filepath-permissions-4813-2)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_18"></a>|@cyli commented|
|-|-|

(In [31134]) Change test to use a 007 permission instead of 001 so test files could delete stuff

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli set owner to @cyli|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_20"></a>|@cyli commented|
|-|-|

(In [31174]) Fix my appalling misunderstanding of stat bitmasks

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_21"></a>|@cyli commented|
|-|-|

(In [31175]) One more test

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_23"></a>|@cyli commented|
|-|-|

[new build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/provide-filepath-permissions-4813-2)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)|@glyph set owner to @glyph<br>@glyph set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_25"></a>|@glyph set owner to @cyli<br>@glyph set status to `new`|
|-|-|

Thanks again for slogging deep into the awful `FilePath` mines to keep the rest
of us supplied with fresh POSIX path abstractions.  It is a shame that so few
people know the arduous work that goes into bringing them their files, but I
will make sure your sacrifice will not be forgotten.

First and foremost, tests apparently fail on Windows.  Sorry you couldn't see
this earlier; I don't know if there's an issue with the buildslaves or what, but
I had to force my own build to actually see results on Windows.

http://buildbot.twistedmatrix.com/builders/winxp32-py2.5/builds/639/steps/select/logs/problems

There are a couple of other issues too, though.  First, review requirements:

  1. `RWX` and `Permissions` are new classes and should therefore be new-style.
     This means inheriting from `object`.
  2. You've implemented `__eq__` and not `__ne__`.  A nice way to get both
     pieces of functionality for free, though, is to use
     `twisted.python.util.FancyEqMixin`.  (Note that you should put `object` at
     the ''end'' of your base class list in `RWX` and `Permissions` to make them
     new-style, as `FancyEqMixin` is unfortunately old-style.)  This indicates a
     deficiency in the test coverage though, because there should be a failing
     test that indicates one `Permissions` or `RWX` is _not un-equal_ to one
     that has a different identity but the same attributes.
  3. `_rwxFromStat` should really have a docstring.  It's private, so it's only
  for maintainers, but jeez it is not entirely obvious what that function does
  or what those parameters are, for someone that isn't really clear on how these
  functions work in the first place :).
  4. Both of the `__str__`s are interesting and perhaps useful, but need to be
  documented.
  5. In the tests, please get rid of the use of the phrases "Make sure that" and
  "Ensure that".  Test docstrings should read like: "`L{foo} returns a value between L{bar} and L{baz}.`", and then the test should go on to demonstrate
  that it does in fact do so.  Also, they should be complete sentences, so
  please put periods at the end.

The following are merely design suggestions, but may impact the way you decide
to deal with the review requirements, so you might want to read them before
responding to the review.

  1. `RWX`'s `__str__` is cute but it would be nice if it had a separate
  `__repr__` which looked something like `RWX(read=True, write=True, execute=False)`, for a little more clarity as to what's being represented.
  2. `RWX` is a bit limiting of a class name, as there are potentially other
  permissions one might have, such as Linux's extended security bits.  If we
  wished to represent such a thing (although I currenty doubt we ever would, who
  knows what the future will bring) this would be the class which would have to
  hold it.
  3. It seems a bit weird to be changing the way 'stat' stuff is imported just
  so the `getattr()` hack can work.  It seems like it might be worthwhile to
  just spell out something a bit more explicit, like a static list or lists
  saying something like:
```
  [[S_IRUSR, S_IWUSR, S_IXUSR],
   [S_IRGRP, S_IWGRP, S_IXGRP],
   [S_IROTH, S_IWOTH, S_IXOTH]]
```
  It seems like there's more code in there now to avoid typing this than this
  actually takes to type.
  4. Apropos of required point 4 above, I am highly suspicious of `__str__`
  methods that output structured formats (like 'rwxr-xr-x') these days; if you
  do str(something) expecting that format, you get a silent and ugly failure if
  'something' isn't the type you think it is.  (For example: if it's `None`.)  I
  think I'd suggest an explicit method named `shorthand()` or something like
  that which returned this documented format of string, and let the `__repr__` I
  suggested in optional point 2 just fill in for the `__str__` as well (which
  python will do automatically if you don't have a `__str__`).
  5. Thanks a lot for the bonus docstrings.  This drew my attention to `isabs`,
  which is a super stupid method, as I am pretty sure that `FilePath` paths are
  _by definition_ absolute, so this will always return True.  You might want
  to note something to that effect.

Again, thanks.  For the most part, this branch looks pretty great, and I'm
looking forward to seeing it land.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_26"></a>|@cyli commented|
|-|-|

(In [31941]) Changed the way stat import works back to the way it used to be.

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_27"></a>|@cyli commented|
|-|-|

(In [31947]) Address non-windows-failure feedback

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_28"></a>|@cyli commented|
|-|-|

(In [31948]) Forgot to commit test

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_29"></a>|@cyli commented|
|-|-|

(In [31951]) Windows execute bit doesn't seem settable via chmod

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_30"></a>|@cyli removed owner|
|-|-|

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/provide-filepath-permissions-4813-2)

Changed branch to conform with feedback, except for the naming of RWX.  I can't really think of a better name that isn't incredibly long - something like TripletPermissionBitsForSinglePermissionClass, which would mean then changing Permissions to something like PermissionsForTheUserGroupAndOtherClasses.

If Linux setuid bits are supported in the future, Permissions can contain a RWXST class instead of an RWX class.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_31"></a>|@thijstriemstra set owner to @cyli|
|-|-|

The `@since` markers introduced in r31133 need to be updated and I don't think they should be added to test-cases. And maybe ReadWriteExecute something?

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_32"></a>|@cyli commented|
|-|-|

(In [32124]) Updated @since markers in filepath, deleted @since markers in test

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)|@cyli removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_34"></a>|@therve set owner to @cyli|
|-|-|

1.
```
+        self.assertTrue(filepath.RWX(True, True, True) !=
+                        filepath.RWX(True, True, False))
+        self.assertTrue(filepath.RWX(True, True, True) != 3)
```

```
+        self.assertFalse(filepath.Permissions(0777) !=
+                         filepath.Permissions(0777))
+        self.assertTrue(filepath.Permissions(0777) !=
+                            filepath.Permissions(0700))
+        self.assertTrue(filepath.Permissions(0777) != 3)
```

You should use assertNotEqual here.

2.
```
+                    chmodString = "0%d%d%d" % (u, g, o)
+                    chmodVal = eval(chmodString)
```

Isn't that the same as int(chmodString, 8)?

3.
```
+                self.failUnlessTrue(getattr(getattr(perm, who), what))
```

Please use asserTrue here.

4.
```
+        self.failUnlessTrue(not self.path.isBlockDevice())
+        self.failUnlessTrue(not self.path.isSocket())
```

And assertFalse here (removing the not).

5. If you feel like it, using assertEqual instead of assertEquals would be nice.

6. The `@since` should say 11.1.

7. The `is*` methods all use the same logic to retrieve the statinfo, it should be factored out.

8. The docstring of isabs changed from `True, always` to something more meaningful, but isn't it still true anyway? I suppose the filepath are always absolute by definition?


There are a bunch of points, but pretty minor. Please merge when you feel like it. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_35"></a>|@cyli commented|
|-|-|

(In [32212]) Change tests to assertEquals, assertFalse, and a new assertNotUnequal method.  Also changed @since from 12.0 to 11.1

refs [#4813](https://github.com/twisted/twisted/issues/4813)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_36"></a>|@cyli commented|
|-|-|

Have addressed (1), (2), (3), and (4).  As part of 1, added a method assertNotUnequal, which tests the `__ne__` method as opposed to `__eq__`.

Filed a ticket ([#5172](https://github.com/twisted/twisted/issues/5172)) to address those, as well as (5), the Twisted coding standard documentation.

Not all of the is* methods are part of this ticket, so I have filed a new ticket to address that:  [#5171](https://github.com/twisted/twisted/issues/5171)

For (8) I think so.  It seems like somewhat of a superfluous method in that case.  Perhaps the path passed to FilePath was meant to be stored at some point, and isabs should have tested whether that was absolute?

At any rate, will attempt to merge.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_37"></a>|@cyli set status to `closed`|
|-|-|

(In [32213]) Merge provide-filepath-permissions-4813-2: filepath permissions, isBlockDevice, and isSocket accessors

Author: cyli
Reviewer: glyph, therve
Fixes: [#4813](https://github.com/twisted/twisted/issues/4813)

Provide an accessor for getting information on filepath permissions, whether the underlying FilePath is a block device, and whether it's a socket.  This is in preparation for deprecating FilePath.statinfo
